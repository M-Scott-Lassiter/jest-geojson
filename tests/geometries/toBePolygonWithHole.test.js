const { goodGeometry, badGeometry, nestedGeometryCollection } = require('./data')

const invalidInputValues = [
    goodGeometry.lineString,
    goodGeometry.multiLineString,
    goodGeometry.multiPoint,
    goodGeometry.multiPolygon,
    goodGeometry.point,
    nestedGeometryCollection,
    badGeometry.lineString,
    badGeometry.multiLineString,
    badGeometry.multiPoint,
    badGeometry.multiPolygon,
    badGeometry.point,
    badGeometry.geometryCollection,
    badGeometry.unrecognized,
    badGeometry.polygon,
    undefined,
    null,
    true,
    false,
    0,
    NaN,
    [
        [
            [0, 0],
            [0, 1],
            [1, 1],
            [1, 0],
            [0, 0]
        ]
    ],
    [[1, 1]],
    { someProp: 'I am not GeoJSON', id: 4 },
    {},
    '',
    'Random Feature',
    JSON.stringify({
        type: 'Polygon',
        coordinates: [
            [
                [100.0, 0.0],
                [101.0, 0.0],
                [101.0, 1.0],
                [100.0, 1.0],
                [100.0, 0.0]
            ],
            [
                [100.8, 0.8],
                [100.8, 0.2],
                [100.2, 0.2],
                [100.2, 0.8],
                [100.8, 0.8]
            ]
        ]
    })
]

describe('Valid Use Cases', () => {
    test('Good Polygon With a Single Hole', () => {
        const testPolygon = {
            type: 'Polygon',
            coordinates: [
                [
                    [100.0, 0.0],
                    [101.0, 0.0],
                    [101.0, 1.0],
                    [100.0, 1.0],
                    [100.0, 0.0]
                ],
                [
                    [100.8, 0.8],
                    [100.8, 0.2],
                    [100.2, 0.2],
                    [100.2, 0.8],
                    [100.8, 0.8]
                ]
            ]
        }
        expect(testPolygon).toBePolygonWithHole()
    })

    test('Good Polygon With Two Holes', () => {
        const testPolygon = {
            type: 'Polygon',
            coordinates: [
                [
                    [100.0, 0.0],
                    [101.0, 0.0],
                    [101.0, 1.0],
                    [100.0, 1.0],
                    [100.0, 0.0]
                ],
                [
                    [100.3, 0.3],
                    [100.3, 0.2],
                    [100.2, 0.2],
                    [100.2, 0.3],
                    [100.3, 0.3]
                ],
                [
                    [100.7, 0.7],
                    [100.7, 0.5],
                    [100.5, 0.5],
                    [100.5, 0.7],
                    [100.7, 0.7]
                ]
            ]
        }
        expect(testPolygon).toBePolygonWithHole()
    })

    test('Good Polygon With 30 Holes', () => {
        const testPolygon = {
            type: 'Polygon',
            coordinates: [
                [
                    [0, 0],
                    [170, 0],
                    [170, 80],
                    [0, 80],
                    [0, 0]
                ]
            ]
        }
        for (let i = 0; i < 30; i++) {
            testPolygon.coordinates.push([
                [i, 1],
                [i, 2],
                [i + 0.5, 2],
                [i + 0.5, 1],
                [i, 1]
            ])
        }
        expect(testPolygon).toBePolygonWithHole()
    })

    test('Hole Has 0 Area', () => {
        const testPolygon = {
            type: 'Polygon',
            coordinates: [
                [
                    [-10, -10],
                    [10, -10],
                    [10, 10],
                    [-10, 10],
                    [-10, -10]
                ],
                [
                    [0, 0],
                    [0, 0],
                    [0, 0],
                    [0, 0]
                ]
            ]
        }
        expect(testPolygon).toBePolygonWithHole()
    })

    test('Hole is Outside Geometry', () => {
        const testPolygon = {
            type: 'Polygon',
            coordinates: [
                [
                    [-10, -10],
                    [10, -10],
                    [10, 10],
                    [-10, 10],
                    [-10, -10]
                ],
                [
                    [-50, 25],
                    [-50, 35],
                    [-45, 30],
                    [-50, 25]
                ]
            ]
        }
        expect(testPolygon).toBePolygonWithHole()
    })

    test('Hole is Wound CounterClockwise', () => {
        const testPolygon = {
            type: 'Polygon',
            coordinates: [
                [
                    [-10, -10],
                    [10, -10],
                    [10, 10],
                    [-10, 10],
                    [-10, -10]
                ],
                [
                    [1, 2],
                    [1, 1],
                    [2, 1],
                    [1, 2]
                ]
            ]
        }
        expect(testPolygon).toBePolygonWithHole()
    })
})

describe('Inalid Use Cases', () => {
    describe('Expect to fail with bad inputs:', () => {
        test.each([...invalidInputValues])('expect(%p).not.toBePolygonWithHole()', (badInput) => {
            expect(badInput).not.toBePolygonWithHole()
        })
    })

    test('Polygon with no hole', () => {
        const polygon = {
            type: 'Polygon',
            coordinates: [
                [
                    [100.0, 0.0],
                    [101.0, 0.0],
                    [101.0, 1.0],
                    [100.0, 1.0],
                    [100.0, 0.0]
                ]
            ]
        }
        expect(polygon).not.toBePolygonWithHole()
    })
})

describe('Error Snapshot Testing. Throws error:', () => {
    test('Valid use case passes', () => {
        expect(() =>
            expect({
                type: 'Polygon',
                coordinates: [
                    [
                        [100.0, 0.0],
                        [101.0, 0.0],
                        [101.0, 1.0],
                        [100.0, 1.0],
                        [100.0, 0.0]
                    ],
                    [
                        [100.8, 0.8],
                        [100.8, 0.2],
                        [100.2, 0.2],
                        [100.2, 0.8],
                        [100.8, 0.8]
                    ]
                ]
            }).not.toBePolygonWithHole()
        ).toThrowErrorMatchingSnapshot()
    })

    test('Invalid input to matcher', () => {
        expect(() => expect(false).toBePolygonWithHole()).toThrowErrorMatchingSnapshot()
    })

    test('Polygon has no holes, expected at least one', () => {
        const testPolygon = {
            type: 'Polygon',
            coordinates: [
                [
                    [100.0, 0.0],
                    [101.0, 0.0],
                    [101.0, 1.0],
                    [100.0, 1.0],
                    [100.0, 0.0]
                ]
            ]
        }
        expect(() => expect(testPolygon).toBePolygonWithHole()).toThrowErrorMatchingSnapshot()
    })
})
