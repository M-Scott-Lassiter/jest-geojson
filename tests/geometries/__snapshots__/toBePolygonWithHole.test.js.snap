// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Snapshot Testing. Throws error: Invalid input to matcher 1`] = `
"[2mexpect([22m[31mGeometryObject[39m[2m).toBePolygonWithHole()[22m

Must have a type property with value 'Polygon'.

Received:  [31mfalse[39m"
`;

exports[`Error Snapshot Testing. Throws error: Polygon has no holes, expected at least one 1`] = `
"[2mexpect([22m[31mGeometryObject[39m[2m).toBePolygonWithHole()[22m

Polygon has no holes, expected at least one.

Received:  [31m{\\"coordinates\\": [[[100, 0], [101, 0], [101, 1], [100, 1], [100, 0]]], \\"type\\": \\"Polygon\\"}[39m"
`;

exports[`Error Snapshot Testing. Throws error: Valid use case passes 1`] = `
"[2mexpect([22m[31mGeometryObject[39m[2m).not.toBePolygonWithHole()[22m

Expected input to not be a valid GeoJSON Polygon geometry with at least one hole.

Received:  [31m{\\"coordinates\\": [[[100, 0], [101, 0], [101, 1], [100, 1], [100, 0]], [[100.8, 0.8], [100.8, 0.2], [100.2, 0.2], [100.2, 0.8], [100.8, 0.8]]], \\"type\\": \\"Polygon\\"}[39m"
`;
